### PR TITLE
Gradle Task support for Code Coverage

### PR DESCRIPTION
**Pull request by nigurr**
With this feature, now gradle task can support Code Coverage feature.

Testing: 
UTs
End to End base scenarios

### CODE REVIEW

**Comment by DavidStaheli** [comment]
Nice!!  You'll also want to bump the version in Tasks/Gradle/task.json to 1.0.**32**.

**Comment by nigurr** [comment]
Fixed that. Please do review :)

https://github.com/Microsoft/vso-agent/pull/273

**Comment by DavidStaheli** [comment]
Looks good to me.  @bryanmacfarlane and @chrisrpatterson, any thoughts about the addition of "string" and "uuid" to package.json?

File: Tasks/Gradle/gradletask.ts [hunk 1]
```diff
@@ -4,88 +4,165 @@ import tl = require('vsts-task-lib/task');
...
+gb.arg(gbTasks);
+gb.arg(ccReportingTask)
+gb.exec()
+    .then(function(code) {
+        publishTestResults(publishJUnitResults);
+        publishCodeCoverage();
+        tl.exit(code);
+    })
+    .fail(function(err) {
+        publishTestResults(publishJUnitResults);
```
Comment (AshwiniChalla) [inline Tasks/Gradle/gradletask.ts]: if publish code coverage failed we are trying to publish test results again.

File: Tasks/Gradle/gradletask.ts [hunk 2]
```diff
@@ -4,88 +4,165 @@ import tl = require('vsts-task-lib/task');
...
 tl.cd(cwd);
 
 var gb = tl.createToolRunner(wrapperScript);
-
-gb.arg(tl.getDelimitedInput('options', ' ', false));
-gb.arg(tl.getDelimitedInput('tasks', ' ', true));
-
-// update JAVA_HOME if user selected specific JDK version or set path manually
+var gbOptions = tl.getDelimitedInput('options', ' ', false);
+var gbTasks = ['clean'];
```
Comment (AshwiniChalla) [inline Tasks/Gradle/gradletask.ts]: Why should we force clean? If user intends to clean he will give it in options.

File: Tasks/Gradle/gradletask.ts [hunk 3]
```diff
@@ -4,88 +4,165 @@ import tl = require('vsts-task-lib/task');
...
+gbTasks = gbTasks.concat(tl.getDelimitedInput('tasks', ' ', true));
 var javaHomeSelection = tl.getInput('javaHomeSelection', true);
-var specifiedJavaHome = null;
+var jdkVersion = tl.getInput('jdkVersion');
+var jdkArchitecture = tl.getInput('jdkArchitecture');
+var buildFolder = tl.getVariable('agent.buildDirectory');
+var publishJUnitResults = tl.getInput('publishJUnitResults');
+var buildTool = "gradle";
+var ccTool = tl.getInput('codeCoverageTool');
+var classFileDirs = tl.getInput('classFilesDirectories');
```
Comment (AshwiniChalla) [inline Tasks/Gradle/gradletask.ts]: I dont see this input from user. where are we getting this from?

File: Tasks/Gradle/gradletask.ts [hunk 4]
```diff
@@ -4,88 +4,165 @@ import tl = require('vsts-task-lib/task');
 import fs = require('fs');
 import path = require('path');
 
+var uuid = require('uuid');
```
Comment (AshwiniChalla) [inline Tasks/Gradle/gradletask.ts]: node-uuid

File: Tasks/Gradle/gradletask.ts [hunk 5]
```diff
@@ -4,88 +4,165 @@ import tl = require('vsts-task-lib/task');
...
     }
 
-    var tp = new tl.TestPublisher("JUnit");
-    tp.publish(matchingTestResultsFiles, false, "", "", "", "");
-  } 
+    var buildProps: { [key: string]: string } = {};
+    buildProps['classfiledirs'] = classFileDirs;
+    buildProps['classfilters'] = classFilters;
+    buildProps['summaryfile'] = summaryFile;
+    buildProps['reportdir'] = reportDir;
```
Comment (AshwiniChalla) [inline Tasks/Gradle/gradletask.ts]: cleaner way is to add all these varables classFileDirs, classFilters, summaryFile, reportDir, isMultimodule to a class and pass the class to this method.

File: Tasks/Gradle/gradletask.ts [hunk 6]
```diff
@@ -4,88 +4,165 @@ import tl = require('vsts-task-lib/task');
...
+    })
+    .fail(function(err) {
+        publishTestResults(publishJUnitResults);
+        publishCodeCoverage();
+        tl.error(err.message);
+        tl.debug('taskRunner fail');
+        tl.exit(1);
+    })
+
+function publishTestResults(publishJUnitResults: string) {
```
Comment (AshwiniChalla) [inline Tasks/Gradle/gradletask.ts]: junit doesnt make sense. change it to publishTestResults and make it bool

File: Tasks/Gradle/gradletask.ts [hunk 7]
```diff
@@ -4,88 +4,165 @@ import tl = require('vsts-task-lib/task');
...
-  } 
+    var buildProps: { [key: string]: string } = {};
+    buildProps['classfiledirs'] = classFileDirs;
+    buildProps['classfilters'] = classFilters;
+    buildProps['summaryfile'] = summaryFile;
+    buildProps['reportdir'] = reportDir;
+    buildProps['buildfile'] = path.join(cwd + "/build.gradle");
+    buildProps['ismultimodule'] = String(isMultiModule);
+
+    //TODO License for all files
```
Comment (AshwiniChalla) [inline Tasks/Gradle/gradletask.ts]: what is this?

File: Tasks/Gradle/gradletask.ts [hunk 8]
```diff
@@ -4,88 +4,165 @@ import tl = require('vsts-task-lib/task');
...
-    tl.exit(code);
-})
-.fail(function(err) {
-    publishTestResults(publishJUnitResults, testResultsFiles);
-    console.error(err.message);
-    tl.debug('taskRunner fail');
-    tl.exit(1);
-})
+function publishCodeCoverage() { 
+    if (codeCoverageOpted) {
```
Comment (AshwiniChalla) [inline Tasks/Gradle/gradletask.ts]: passing this as argument to function is a more clean way and more understandable.

File: Tests/lib/vsts-task-lib/task.ts [hunk 9]
```diff
@@ -533,6 +538,7 @@ export function globFirst(pattern: string): string {
 // Exec convenience wrapper
 //-----------------------------------------------------
 export function exec(tool: string, args: any, options?: trm.IExecOptions): Q.Promise<number> {
+    debug('tool ' + tool);
```
Comment (AshwiniChalla) [inline Tests/lib/vsts-task-lib/task.ts]: do we need this?

File: Tests/L0/Gradle/_suite.ts [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        var mockFile = path.join(__dirname, mockResponse);
+        var mockData = fs.readFileSync(mockFile, 'utf-8');
+        fs.unlinkSync(mockFile);
+        fs.writeFileSync(mockFile, s(mockData).replaceAll('${buildfolder}', temp));
+    });
+
+    after(function() {
+    });
+
+    it('Gradle build without code coverage', (done) => {
```
Comment (AshwiniChalla) [inline Tests/L0/Gradle/_suite.ts]: Hope all tests are ported from windows agent implementation tests

**Comment by AshwiniChalla** [comment]
approved with comments. 
Make sure you checkin this only after vsts-task lib and vso-agent are released with your changes
or else make sure that vsts-task lib and vso-agent are released before next deployment.

File: Tests/lib/vsts-task-lib/task.ts (hunk 9 above)
Comment (nigurr) [inline Tests/lib/vsts-task-lib/task.ts]: it's to debug which tool is getting called

File: Tasks/Gradle/gradletask.ts (hunk 6 above)
Comment (nigurr) [inline Tasks/Gradle/gradletask.ts]: all input variables are coming as string.

File: Tasks/Gradle/gradletask.ts [hunk 11]
```diff
@@ -4,88 +4,163 @@ import tl = require('vsts-task-lib/task');
...
+var buildFolder = tl.getVariable('agent.buildDirectory');
+var publishJUnitResults = tl.getInput('publishJUnitResults');
+var buildTool = "gradle";
+var ccTool = tl.getInput('codeCoverageTool');
+var classFileDirs = tl.getInput('classFilesDirectories');
+var classFilter = tl.getInput('classFilter');
+var summaryFile = null;
+var reportDir = null;
+var ccReportingTask = "";
+var codeCoverageOpted = (typeof ccTool != "undefined" && ccTool && ccTool.toLowerCase() != 'none');
```
Comment (AshwiniChalla) [inline Tasks/Gradle/gradletask.ts]: isCodeCoverageOpted

**Pull request closed without merging**